### ISSUE TITLE
XA Transactions are running on multiple threads

### ISSUE THREAD

**Issue by wkonecny** [hazelcast/hazelcast#3385]
We are trying to use the Hazelcast Resource Adapter with a Oracle Weblogic container. When a XA transaction started it gave an unexpected Exception.

After some research, it turns out that the transactions _are_ running on multiple threads. However, Hazelcast explicitly checks for multiple threads and throws an Exception when this is the case. This causes a rollback to occur. What is the reasoning for not allowing multithreaded transactions?

`com.hazelcast.transaction.impl.TransactionImpl.java:177`
```java
private void checkThread() {
    if (checkThreadAccess && threadId != null && threadId.longValue() != Thread.currentThread().getId()) {
        throw new IllegalStateException("Transaction cannot span multiple threads!");
    }
}
```

Additional Thead info

| State | Thread no. Run 1 | Thread no. Run 2 |
| ------------- | ------------- | ------------- |
| begin | 15 | 15 |
| prepare | 53 | 64 |
| commit |  45 | 52 |

**Comment by juanavelez** [comment]
Any updates? We are really interested in using HZ within our JBoss EAP transactions but this is a no-no if it is not fixed.

**Comment by jerrinot** [comment]
@gurbuzali This one is gonna be fixed in 3.5. Is it correct?

**Comment by gurbuzali** [comment]
should be fixed in EA

**Comment by jerrinot** [comment]
This is fixed in the current master and it'll be included in Hazelcast 3.5.
